motionplan/nearestNeighbor.go

```diff
@@ -138,43 +124,30 @@ func (nm *neighborManager) nnWorker(ctx context.Context, planOpts *plannerOption
 	var best node
 	bestDist := math.Inf(1)
 
-	for {
-		select {
-		case <-ctx.Done():
-			nm.neighbors <- nil
-			return
-		default:
+	for candidate := range nm.nnKeys {
+		seg := &Segment{
+			StartConfiguration: nm.seedPos.Q(),
+			EndConfiguration:   candidate.Q(),
+		}
+		if pose := nm.seedPos.Pose(); pose != nil {
+			seg.StartPosition = pose
+		}
+		if pose := candidate.Pose(); pose != nil {
+			seg.EndPosition = pose
+		}
+		dist := planOpts.DistanceFunc(seg)
+		if dist < bestDist {
+			bestDist = dist
+			best = candidate
 		}
 
 		select {
-		case k := <-nm.nnKeys:
-			if k != nil {
-				nm.nnLock.RLock()
-				seg := &Segment{
-					StartConfiguration: nm.seedPos.Q(),
-					EndConfiguration:   k.Q(),
-				}
-				if pose := nm.seedPos.Pose(); pose != nil {
-					seg.StartPosition = pose
-				}
-				if pose := k.Pose(); pose != nil {
-					seg.EndPosition = pose
-				}
-				dist := planOpts.DistanceFunc(seg)
-				nm.nnLock.RUnlock()
-				if dist < bestDist {
-					bestDist = dist
-					best = k
-				}
-			}
+		case <-ctx.Done():
+			nm.neighbors <- &neighbor{bestDist, best}
```

Comment:
The only functional change of this PR. In the old code -- if an `nnWorker` saw the context canceled, it would refuse to report its progress (returned a `nil`). And the nearest neighbor method gathering all the results from workers would simply ignore that `nil`. With this PR though I flipped that.

I was thinking about the question of what are the possible outcomes with the return values from `nnWorker`s falling into these 3 buckets:
1. All inputs are consumed, each nnWorker returns its version of the best answer.
2. The context is interrupted -- every nnWorker returns nil.
3. The context is interrupted -- some workers return their best candidate and some return nil.

I argue that (3) was only possible in an unhelpful way. I claim in that case, between all of the workers, every input was looked at. And the best value would have been returned by the workers that instead return `nil`. I'm not going to write down a proof, but I'm happy to go over that argument in more rigorous detail offline

I think the more interesting question is -- if the context is interrupted, should we return the "best so far" answer, or nothing to the top-level caller (and give the caller a way to distinguish between receiving the true answer vs our best approximation).

If we want to return nothing in that case, I'm happy to make a change.

Reply:
I would prefer to return `nil` in this case rather than "best so far". I'm having trouble thinking of a case where "best so far", which functionally could be any random node, is useful.

Reply:
So I think you're advocating that if a context is canceled, we should never return a result to the top-level caller. The caller can only get the single right answer or no answer at all. To achieve that, having this return nil is not sufficient (especially if we can't convince ourselves of a partial result -- some workers return nil and some return a real value). Consider the following:
```
Target: 10
Inputs: [7, 8, 9, 11]

| Worker 1                   | Worker 2                           |
|----------------------------+------------------------------------|
| Sees 7 - Best: 7 (diff: 3) |                                    |
| Sees 8 - Best: 8 (diff: 2) |                                    |
|                            | Sees 9 - Best: 9 (diff 1)          |
|                            | Sees 11 - Best: 9                  |
| Returns 8                  |                                    |
|                            | Sees context canceled, returns nil |
```
In this case we actually return a sub-optimal answer

If we want to return no answer at all to the top-level caller, what I would do instead of returning nil is to have the accumulator check the context, e.g:
```
diff --git a/motionplan/nearestNeighbor.go b/motionplan/nearestNeighbor.go
index 351170c6..b11c09b5 100644
--- a/motionplan/nearestNeighbor.go
+++ b/motionplan/nearestNeighbor.go
@@ -131,6 +131,9 @@ func (nm *neighborManager) parallelNearestNeighbor(
                default:
                }
        }
+       if ctx.Err() != nil {
+               return nil
+       }
        return best
 }
```

Or it might be more robust to change the accumulator to check for nils (as it does today) and when seeing a nil, bubbles up a nil/error, e.g:
```
diff --git a/motionplan/nearestNeighbor.go b/motionplan/nearestNeighbor.go
index 351170c6..d6facfb9 100644
--- a/motionplan/nearestNeighbor.go
+++ b/motionplan/nearestNeighbor.go
@@ -121,12 +121,13 @@ func (nm *neighborManager) parallelNearestNeighbor(
                select {
                case nn := <-nm.neighbors:
                        returned++
-                       if nn != nil {
+                       if nn == nil {
                                // nn will be nil if the ctx is cancelled
-                               if nn.dist < bestDist {
-                                       bestDist = nn.dist
-                                       best = nn.node
-                               }
+                               return nil
+                       }
+                       if nn.dist < bestDist {
+                               bestDist = nn.dist
+                               best = nn.node
                        }
                default:
                }
```

Reply:
> So I think you're advocating that if a context is canceled, we should never return a result to the top-level caller. The caller can only get the single right answer or no answer at all.

This is correct. The reason being that the case where workers return a mix of `nil` and true values should only be a vanishingly small fraction of cancellation instances, requiring the cancellation to happen during the runtime of the very last iteration of workers. It is far more likely that if a cancellation happens during NN runtime, it happens while some subset of `nnKeys` remains unevaluated. In this case the "best" value returned is essentially random.

I think the most robust solution is to have the accumulator check for `nil` as in your example immediately above and bubble that up.

Reply:
Done. That makes sense to me. I put this version up because it was closer to what the original code was doing (making a best effort) and I couldn't tell if that best effort was intentional or a "bug".